sql: adjust physical planning heuristics around small joins

yuzefovich opened the pull request:
This PR adjusts the physical planner heuristics so that we no longer force the plan distribution in `distsql=auto` mode when we have a hash or a merge join with at least one equality column (i.e. non-cross join) when we have "small" inputs (less than 1k rows combined). See each commit for details.

Epic: None

yuzefovich committed:
sql: fix physical planner heuristic in edge case with cross joins

Long time ago in a74611e45e01ebcef42b38ecfdcfd57f3e9d2331 we introduced
a change so that when we merge two plans when performing a cross join
and both plans have a single result router (i.e. a single processor), we
would plan the cross join on the same node as the _left_ input. This was
done under the assumption that the left input is processed first by the
joiner, but this assumption has changed since then (long time ago too),
so we should be choosing the _right_ input. The impact seems very minor
though, just something I noticed while working around this area.

Release note: None

yuzefovich committed:
sql: adjust physical planning heuristics around small joins

Previously, whenever we had a hash or a merge join with at least one
equality column (i.e. not a cross join), we would force the distribution
of the plan. This can be suboptimal when we need to process a small
number of rows. This commit makes the heuristic a bit more configurable:
we now will choose to distribute only when performing a "large" join,
where "large" is defined as having both inputs produce at least 1k rows
(configured via new `distribute_join_row_count_threshold` session
variable). If both inputs don't have stats available, we fall back to
the old behavior of distributing the plan (which seems like a safer
option).

Note the following operations still force the plan distribution
unconditionally:
- table statistics
- inverted filter (when we have a union of inverted spans)
- inverted join
- window functions with at least one window frame with PARTITION BY
clause
- zigzag join.

I think there are cases where we don't want to force the distribution
because of these operations, but they seem less of a concern, so I left
a few TODOs. Ideally, eventually we'll have the optimizer make the
decision.

Release note (sql change): DistSQL physical planning decisions under
`distsql=auto` mode have been adjusted in the following manner:
presence of the hash or merge join no longer forces the plan to be
distributed. Namely, we might not choose to distribute the plan if both
inputs to the join are expected produce small number of rows (less than
1k combined by default, configurable via
`distribute_join_row_count_threshold` session variable).

yuzefovich committed:
sql: add some logging to physical planning

This commit introduces some logging, hidden behind a verbosity level, to
the physical planning whenever we choose `shouldDistribute`
recommendation. This will show up in the trace and should be helpful to
understand why we decided to distribute a particular query.

Release note: None

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice tracing! That's going to come in handy! Thanks for doing this.

> Reviewed 2 of 2 files at r1, 13 of 13 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137562#-:-OELnyD42_AgmZnpMcG6:bc24j2c)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @yuzefovich)

___
*[`-- commits` line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/137562#-OELpuoK3chlDs_Z4zWP:-OELpuoK3chlDs_Z4zWQ:b-uu7rec):*
Doesn't have to be part of this PR, but it would be nice to also have the tracing if we choose `cannotDistribute`, for the opposite reason (understanding why we didn't distribute a particular query).
___
*[`pkg/sql/distsql_physical_planner.go` line 612 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137562#-OELkByu0AgZV7jCPxb4:-OELkByv0Ngptis3AARh:b-ziiaw2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2db3126b8ff10e351c6d5748555d72cf712f1ce/pkg/sql/distsql_physical_planner.go#L612)):*
> ```Go
> 				// distributing.
> 				rec = rec.compose(shouldDistribute)
> 			} else if n.estimatedLeftRowCount+n.estimatedRightRowCount >= sd.DistributeJoinRowCountThreshold {
> ```

I've said it before, and I'll say it again: with all these row count thresholds, I feel like distribution is really begging to be a cost-based optimizer decision.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: ensure that cannotDistribute recommendation always has an error

We log the error into the trace which is helpful to understand why
a particular query couldn't be distributed.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137562)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @michae2)

___
*[`-- commits` line 55 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/137562#-OELpuoK3chlDs_Z4zWP:-OEM2BvgEd48SYX13TVa:b-7snnv1):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Doesn't have to be part of this PR, but it would be nice to also have the tracing if we choose `cannotDistribute`, for the opposite reason (understanding why we didn't distribute a particular query).
</blockquote></details>

Good point. I initially thought about doing that but then recalled that in this case we return an error for `checkSupportForPlanNode` that is already logged (done in `getPlanDistribution`).

However, your suggestion prompted me to look over all spots where we use `cannotDistribute`, and there were some where we returned `nil` error, so I added another commit to fix those and a simple test-only assertion to always have an error with `cannotDistribute` recommendation.
___
*[`pkg/sql/distsql_physical_planner.go` line 612 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/137562#-OELkByu0AgZV7jCPxb4:-OEM5KfgA0CFlH_oG6rc:b-hs02mg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2db3126b8ff10e351c6d5748555d72cf712f1ce/pkg/sql/distsql_physical_planner.go#L612)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I've said it before, and I'll say it again: with all these row count thresholds, I feel like distribution is really begging to be a cost-based optimizer decision.
</blockquote></details>

Definitely agreed.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/137562#-:-OENLiOzE8ep8VLcvAB7:b-oduykh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner and @michae2)



<!-- Sent from Reviewable.io -->
